Allow moderating values on behalf of instructors #4274

kennho opened the pull request:
Fixes #4274

damithc commented on src/main/java/teammates/ui/controller/InstructorEditInstructorFeedbackSaveAction.java:
```diff
@@ -0,0 +1,187 @@
+                        "Feedback session [" + feedbackSessionName + 
+                        "] question [" + questionAttributes.getId() + "] is not accessible to instructor ["+ instructor.email + "]");
+            }
+        }
```
Body of the `for` loop is rather long. Increase SLAP?

damithc commented:
Remember to write an `*ActionTest` class for new tests.

LowWeiLin commented on src/main/java/teammates/ui/controller/FeedbackSubmissionEditPageData.java:
```diff
@@ -151,6 +151,10 @@ public void setPreviewInstructor(InstructorAttributes previewInstructor) {
         this.registerMessage = registerMessage;
     }
+    
+    public void setSubmitAction(String submitAction) {
```
Instead of a setter, can you have an abstract function that the child classes implement?

kennho replied:
abstract function inside PageData.java?

LowWeiLin replied:
Sorry ignore this comment.

LowWeiLin commented on src/main/java/teammates/ui/controller/InstructorEditInstructorFeedbackPageAction.java:
```diff
@@ -0,0 +1,105 @@
+    protected ActionResult execute() throws EntityDoesNotExistException {
+        String courseId = getRequestParamValue(Const.ParamsNames.COURSE_ID); 
+        String feedbackSessionName = getRequestParamValue(Const.ParamsNames.FEEDBACK_SESSION_NAME);
+        String moderatedEntityIdentifier = getRequestParamValue(Const.ParamsNames.FEEDBACK_SESSION_MODERATED_STUDENT);
```
moderatedEntityIdentifier is not a very good variable name. What entity is it? FEEDBACK_SESSION_MODERATED_STUDENT <- should this be student or instructor?

LowWeiLin commented on src/main/java/teammates/ui/controller/InstructorEditInstructorFeedbackPageAction.java:
```diff
@@ -0,0 +1,105 @@
+        String courseId = getRequestParamValue(Const.ParamsNames.COURSE_ID); 
+        String feedbackSessionName = getRequestParamValue(Const.ParamsNames.FEEDBACK_SESSION_NAME);
+        String moderatedEntityIdentifier = getRequestParamValue(Const.ParamsNames.FEEDBACK_SESSION_MODERATED_STUDENT);
+        String moderatedQuestionNumber = getRequestParamValue("moderatedquestion");
```
magic string here, use a constant instead

LowWeiLin commented on src/main/java/teammates/ui/controller/InstructorEditInstructorFeedbackPageAction.java:
```diff
@@ -0,0 +1,105 @@
+        new GateKeeper().verifyAccessible(
+                logic.getInstructorForGoogleId(courseId, account.googleId), 
+                logic.getFeedbackSession(feedbackSessionName, courseId),
+                false, Const.ParamsNames.INSTRUCTOR_PERMISSION_MODIFY_SESSION);
```
Is this the permission used to allow moderation of student responses? We may need a separate permission...

LowWeiLin commented on src/main/java/teammates/ui/controller/InstructorEditInstructorFeedbackPageAction.java:
```diff
@@ -0,0 +1,105 @@
+                    + ".");
+        }
+
+        FeedbackSubmissionEditPageData data = new FeedbackSubmissionEditPageData(account, student);
```
Where is this student variable from?

kennho replied:
The student variable comes from the parent Action.java
I think it is used to track unregistered students  @LowWeiLin

LowWeiLin replied:
InstructorEditInstructorFeedbackPageAction does not involve students at all right? It seems strange that all FeedbackSubmissionEditPageData needs student as a parameter. Is it possible to refactor it?

kennho replied:
yup I can add another constructor that takes in only the account.

LowWeiLin commented on src/main/java/teammates/ui/controller/InstructorEditInstructorFeedbackPageAction.java:
```diff
@@ -0,0 +1,105 @@
+                        "Session Name: " + feedbackSessionName + "<br>" +
+                        "Course ID: " + courseId;
+        
+        data.init("", "", courseId);
```
This does not make sense just by looking at the function call... we should refactor this.

LowWeiLin commented on src/main/java/teammates/ui/controller/InstructorEditInstructorFeedbackPageAction.java:
```diff
@@ -0,0 +1,105 @@
+     * Removes question from the bundle if the question has givers, recipients or responses that are anonymous to instructors.
+     * @param bundle
+     */
+    private boolean hideQuestionsWithAnonymousResponses(FeedbackSessionQuestionsBundle bundle) {
```
This function seems to be duplicated from InstructorEditStudentFeedbackPageAction. Maybe we can find a better place for it so that there will not be repeated code.

LowWeiLin commented on src/main/java/teammates/ui/controller/InstructorEditInstructorFeedbackSaveAction.java:
```diff
@@ -0,0 +1,195 @@
+        new GateKeeper().verifyAccessible(instructor,
+                session,
+                false, 
+                Const.ParamsNames.INSTRUCTOR_PERMISSION_MODIFY_SESSION_COMMENT_IN_SECTIONS);
```
Why is permission for comments used..?

kennho replied:
Hmm. I was looking at the permission when instructors are moderating students and it uses this for saving of the moderated values and loading of the feedback form. Should I add new permission for instructor to moderate students and instructor to moderate instructor because this also doesn't seem like the permission should used when instructor moderates student. Should all these be done in a separate PR? Or I just include them here? @LowWeiLin

LowWeiLin replied:
Maybe we can use INSTRUCTOR_PERMISSION_MODIFY_SESSION first, and create a separate issue for a new permission.

kennho replied:
okay will make the necessary changes

LowWeiLin commented on src/main/java/teammates/ui/controller/InstructorEditInstructorFeedbackSaveAction.java:
```diff
@@ -0,0 +1,195 @@
+        String moderatedInstructorEmail = getRequestParamValue(Const.ParamsNames.FEEDBACK_SESSION_MODERATED_INSTRUCTOR);
+        Assumption.assertPostParamNotNull(Const.ParamsNames.FEEDBACK_SESSION_MODERATED_INSTRUCTOR, moderatedInstructorEmail);
+
+        moderatedInstructor = logic.getInstructorForEmail(courseId, moderatedInstructorEmail);
```
Need to check if instructor exists?

LowWeiLin commented on src/main/java/teammates/ui/controller/InstructorEditInstructorFeedbackSaveAction.java:
```diff
@@ -0,0 +1,195 @@
+                statusToUser.add(new StatusMessage("The feedback session or questions may have changed while you were submitting. "
+                                                + "Please check your responses to make sure they are saved correctly.", StatusMessageColor.WARNING));
+                isError = true;
+                log.warning("Question not found. (deleted or invalid id passed?) id: "+ questionId + " index: " + questionIndx);
```
This warning message lacks details to be useful. Should also include the course and feedback session.

LowWeiLin commented on src/main/java/teammates/ui/controller/InstructorEditInstructorFeedbackSaveAction.java:
```diff
@@ -0,0 +1,195 @@
+        try {
+            logic.addInstructorRespondant(getUserEmailForCourse(), feedbackSessionName, courseId);
+        } catch (InvalidParametersException | EntityDoesNotExistException e) {
+            log.severe("Fail to append instructor respondant");
```
Failed*
Need to add more details to the log. e.g. course, fs, stacktrace...

LowWeiLin commented on src/main/java/teammates/ui/controller/InstructorEditInstructorFeedbackSaveAction.java:
```diff
@@ -0,0 +1,195 @@
+        try {
+            logic.deleteInstructorRespondant(getUserEmailForCourse(), feedbackSessionName, courseId);
+        } catch (InvalidParametersException | EntityDoesNotExistException e) {
+            log.severe("Fail to remove instructor respondant");
```
Same as above comment

LowWeiLin commented on src/main/java/teammates/ui/controller/InstructorEditInstructorFeedbackSaveAction.java:
```diff
@@ -0,0 +1,195 @@
+        
+        result.responseParams.put(Const.ParamsNames.COURSE_ID, moderatedInstructor.courseId);
+        result.responseParams.put(Const.ParamsNames.FEEDBACK_SESSION_NAME, feedbackSessionName);
+        result.responseParams.put(Const.ParamsNames.FEEDBACK_SESSION_MODERATED_STUDENT, moderatedInstructor.email);
```
Creating a new constant FEEDBACK_SESSION_MODERATED_INSTRUCTOR may be more appropriate.

LowWeiLin commented on src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java:
```diff
@@ -516,12 +516,13 @@ private InstructorFeedbackResultsGroupByParticipantPanel buildGroupByParticipant
+                boolean isInstructor = bundle.roster.getInstructorForEmail(secondaryParticipantIdentifier) != null;
                 String sectionName = bundle.getSectionFromRoster(secondaryParticipantIdentifier);
-                boolean isAllowedToModerate = (isStudent || isVisibleTeam) 
+                boolean isAllowedToModerate = (isStudent || isVisibleTeam || isInstructor) 
```
What is the logic for who is allowed to moderate? Extract the condition out into a function?
Unless i'm mistaken, a person in the course is either student or instructor .. is there a point to isStudent || isInstructor when it will always return true?

LowWeiLin commented on src/main/java/teammates/ui/template/InstructorFeedbackResultsModerationButton.java:
```diff
@@ -11,11 +11,12 @@
                              String giverIdentifier,
                              String courseId, String feedbackSessionName, FeedbackQuestionAttributes question,
-                             String buttonText) {
+                             String buttonText, String moderateFeedbackLink) {
```
moderateFeedbackLink or moderateResponseLink?

LowWeiLin commented:
@kennho Some comments added, remember to add tests as mentioned by @damithc

kennho commented:
updated with suggested changes @LowWeiLin

kennho commented:
Updated with tests @LowWeiLin

thyageshm commented:
@LowWeiLin ready for review

kanghj commented:
@LowWeiLin remember to review this

kennho commented:
Hello @LowWeiLin possible to review this soon? Sorry for the trouble!

LowWeiLin commented on src/test/java/teammates/test/cases/ui/pagedata/FeedbackSubmissionEditPageDataTest.java:
```diff
@@ -78,9 +78,6 @@ public void testAll() {
         
-        assertEquals("/page/studentFeedbackSubmissionEditSave", pageData.getSubmitAction());
-        assertEquals("/page/studentFeedbackQuestionSubmissionEditSave", pageData.getSubmitActionQuestion());
-        
```
Instead of removing the tests, can they be modified?

LowWeiLin commented on src/main/java/teammates/ui/controller/FeedbackSubmissionEditPageData.java:
```diff
@@ -37,6 +42,10 @@ public FeedbackSubmissionEditPageData(AccountAttributes account, StudentAttribut
         isHeaderHidden = false;        
     }
+    
+    public void init() {
```
The function name is not very informative. Add comments for this? And if this is really done for initialization, why is it not in the constructor?

LowWeiLin commented on src/test/java/teammates/test/cases/ui/browsertests/InstructorFeedbackResultsPageUiTest.java:
```diff
@@ -167,7 +167,14 @@ public void testModerateResponsesButton() {
+        
+        resultsPage = loginToInstructorFeedbackResultsPage("CFResultsUiT.instr", "Session with Instructors as Givers");
+        resultsPage.displayByQuestion();
+        ThreadHelper.waitFor(2000);
```
Try to avoid using a waitFor, wait on the page load or element load instead.

LowWeiLin commented on src/test/java/teammates/test/cases/ui/browsertests/InstructorEditStudentFeedbackPageUiTest.java:
```diff
@@ -117,7 +117,7 @@ private FeedbackSubmitPage loginToInstructorEditStudentFeedbackPage(
                 withCourseId(testData.feedbackSessions.get(fsName).courseId).
                 withSessionName(testData.feedbackSessions.get(fsName).feedbackSessionName).
-                withParam(Const.ParamsNames.FEEDBACK_SESSION_MODERATED_STUDENT, moderatedStudentEmail);
+                withParam(Const.ParamsNames.FEEDBACK_SESSION_MODERATED_PERSON, moderatedStudentEmail);
```
No InstructorEditInstructorFeedbackPageUiTest? but i guess its fine as there are no UI stuff.

kennho replied:
Yeah I didn't add as there are no new UI stuff.

kennho commented:
Updated with suggested changes @LowWeiLin 
I have added back the constructor to take in student as I realised it might need it if the student is a unregistered student.

kennho commented:
Hello @Low Wei Lin possible to review this soon? Must get merged by tmr. Or assign someone else to review?

damithc commented on src/main/java/teammates/common/datatransfer/FeedbackSessionQuestionsBundle.java:
```diff
@@ -105,7 +105,27 @@ public int compare(Map.Entry<String, String> o1, Map.Entry<String, String> o2) {
     }
+    
+    /**
+     * Removes question from the bundle if the question has givers, recipients or responses that are anonymous to instructors.
```
hidden, not anonymous

damithc replied:
Actually, it should be `anonymous or hidden`?

kennho replied:
Actually is both. It's either the giver or recipient is anonymous or the response is hidden @damithc.  Will update the comments and method name?

damithc commented on src/main/java/teammates/ui/controller/InstructorEditInstructorFeedbackPageAction.java:
```diff
@@ -0,0 +1,81 @@
+
+        if (moderatedQuestionNumber != null) {
+            data.setModeratedQuestion(moderatedQuestionNumber);
+        }
```
@kanghj is this another place where question id should be used instead?

kanghj replied:
Yes, the questionId should be used here. 

@kennho, can you check if it will require a large change if we try to use feedbackQuestionId here instead of question number?

kennho replied:
So I will have to create a new issue for moderating students first? @kanghj

kanghj replied:
Yes, that would be better.

kennho replied:
Okays. What exactly is the issue uh? Use feedbackQuestionId instead of moderatedQuestionNumber? @kanghj

kanghj replied:
Yup, use feedbackQuestionId instead of questionNumber to identify which question is getting moderated

damithc commented:
Waiting for small change and clarification.

kennho commented:
Updated with suggested changes @LowWeiLin @kanghj @damithc

damithc commented:
My comment was concerning the method comment, not the method name.

damithc commented on src/main/java/teammates/common/datatransfer/FeedbackSessionQuestionsBundle.java:
```diff
@@ -105,11 +105,12 @@ public int compare(Map.Entry<String, String> o1, Map.Entry<String, String> o2) {
+     * or responses that are hidden from the instructors.
      */
-    public void hideQuestionsWithAnonymousResponses() {
+    public void hideQuestionsWithHiddenResponsesOrAnonymousRecipientOrGiver() {
```
`hideUnmoderatableQuestions`?

kennho commented:
Updated with suggested changes @damithc

damithc commented:
As mentioned twice, you need to update the method comment too.

kennho commented on src/main/java/teammates/common/datatransfer/FeedbackSessionQuestionsBundle.java:
```diff
@@ -108,9 +108,9 @@ public int compare(Map.Entry<String, String> o1, Map.Entry<String, String> o2) {
      * Removes question from the bundle if the question has givers or recipients that are anonymous to the instructor
-     * or responses that are hidden from the instructors.
+     * or responses that are hidden from the instructor.
      */
```
I have already updated the method comments. @damithc is this okay?

kennho commented:
I have already updated the comments of the method to 
```
Removes question from the bundle if the question has givers or recipients that are anonymous
to the instructor or responses that are hidden from the instructor.
``` 
@damithc

damithc commented:
Sorry I didn't notice. Yup, the new comment works.

kennho commented:
Updated to use question id instead of question number @LowWeiLin @kanghj @damithc

kanghj commented on src/main/java/teammates/common/util/Const.java:
```diff
@@ -444,7 +444,9 @@ public static final String getjQueryUiFilePath(boolean isDevServer) {
         public static final String FEEDBACK_SESSION_INSTRUCTIONS = "instructions";
-        public static final String FEEDBACK_SESSION_MODERATED_STUDENT = "moderatedstudent";
+        public static final String FEEDBACK_SESSION_MODERATED_PERSON = "moderatedperson";
+        public static final String FEEDBACK_SESSION_MODERATED_QUESTION = "moderatedquestion";
```
@kennho, We can remove this constant

kennho commented:
Updated with suggested changes @kanghj

kanghj commented on src/main/java/teammates/ui/controller/InstructorEditInstructorFeedbackSaveAction.java:
```diff
@@ -0,0 +1,205 @@
+        moderatedInstructor = logic.getInstructorForEmail(courseId, moderatedInstructorEmail);
+        
+        // If the instructor doesn't exist
+        if (moderatedInstructor == null) {
```
Just to clarify, why is this check needed?

kennho replied:
Actually I couldn't really remember why I did it since it was quite some time ago. I think I probably referenced some similar places in the code then did it. Should I remove it? @kanghj

kanghj replied:
Hmm, I just wanted to check. I don't think it is required, but it should be fine to have it. I think the difference will be whether or not the user reaches the entityNotFound page or our generic error page?

kennho replied:
Yeah. I realised why I added it. Because it was suggested to add it @kanghj 
![screen shot 2016-04-10 at 7 10 43 pm](https://cloud.githubusercontent.com/assets/10682388/14409805/2db3eb10-ff50-11e5-8f69-1a869a861d3c.png)

kanghj commented on src/test/java/teammates/test/cases/ui/InstructorEditInstructorFeedbackPageActionTest.java:
```diff
@@ -0,0 +1,125 @@
+        
+        try {
+            editInstructorFPAction = getAction(submissionParams);
+            editInstructorFPAction.executeAndPostProcess();
```
This is missing a <code>signalFailureToDetectException</code>

kanghj commented on src/test/java/teammates/test/cases/ui/InstructorEditInstructorFeedbackPageActionTest.java:
```diff
@@ -0,0 +1,125 @@
+        
+        try {
+            editInstructorFPAction = getAction(submissionParams);
+            editInstructorFPAction.executeAndPostProcess();
```
Likewise here. Missing signalFailureToDetectException

kanghj commented on src/test/java/teammates/test/cases/ui/InstructorEditInstructorFeedbackSaveActionTest.java:
```diff
@@ -0,0 +1,419 @@
+        
+        try {
+            editInstructorFSAction = getAction(submissionParams);
+            editInstructorFSAction.executeAndPostProcess();
```
Likewise here, missing <code>signalFailureToDetectException</code>

kanghj commented on src/test/java/teammates/test/cases/ui/browsertests/InstructorFeedbackResultsPageUiTest.java:
```diff
@@ -156,7 +156,7 @@ public void testModerateResponsesButton() {
         resultsPage = loginToInstructorFeedbackResultsPage("CFResultsUiT.instr", "Open Session");
         resultsPage.displayByQuestion();
-        ThreadHelper.waitFor(2000);
+        resultsPage.waitForPageToLoad();
```
Thanks for removing this! But I think <code>displayByQuestion</code> already performs a waitForPageToLoad?

kanghj commented on src/test/resources/data/InstructorEditInstructorFeedbackPageTest.json:
```diff
@@ -0,0 +1,466 @@
+        "IEIFPTCourse":{
+            "id":"IEIFPTCourse",
+            "name":"IEIFPTCourse",
+            "isArchived":false
```
I think you can remove the <code>isArchived</code> field here

kennho commented:
Updated with suggested changes @kanghj

kanghj commented on src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java:
```diff
@@ -514,16 +514,10 @@ private InstructorFeedbackResultsGroupByParticipantPanel buildGroupByParticipant
-                boolean isShowingModerationButton = (isStudent || isVisibleTeam) 
-                                                 && isAllowedToModerate;
-                secondaryParticipantPanel.setModerationButton(isShowingModerationButton
+                boolean isAllowedToModerate = isAllowedToModerate(instructor, sectionName, feedbackSessionName);
```
? What happened to the check that the team is visible?

kennho replied:
Hmm.. Previously it was 
`isStudent || isVisibleTeam`. 

After allowing instructor to get moderated too, it became 
`isStudent || isVisibleTeam || isInstructor`

It seems like this condition will always be true? @kanghj

kanghj replied:
Perhaps the original variable names there were not good. But if a team is anonymous, then it would be false. <code>isStudent || isVisibleTeam || isInstructor</code> looks more correct

kanghj replied:
Ok, it doesn't matter actually. You can remove this check since its duplicated. <code>buildModerationButtonForGiver</code> will have the correct check.

kennho replied:
Oh okay. So I will just leave it as it is then? @kanghj

kanghj replied:
Yes, this should be fine

kennho replied:
okay

kanghj commented:
I have no other comments. @damithc @LowWeiLin

kanghj commented:
I'll just put this to toMerge since @LowWeiLin already reviewed it previously

damithc commented on src/test/resources/data/InstructorEditInstructorFeedbackPageTest.json:
```diff
@@ -0,0 +1,465 @@
+            "role":"Co-owner",
+            "isInstructor":true,
+            "displayedName":"Instructor",
+            "instructorPrivilegesAsText": "{\n  \"courseLevel\": {\n    \"canviewstudentinsection\": true,\n    \"cangivecommentinsection\": true,\n    \"cansubmitsessioninsection\": true,\n    \"canmodifysessioncommentinsection\": true,\n    \"canmodifycommentinsection\": true,\n    \"canmodifycourse\": true,\n    \"canviewsessioninsection\": true,\n    \"canmodifysession\": true,\n    \"canview ...
```
Didn't Yuxuan merged a PR that affects this field recently? I think it is no longer a string. In any case, the line is too long.

kennho replied:
okay will make the necessary changes

kanghj replied:
@kennho, You will have to make changes again, due to 9c031c4ec

kennho replied:
Okay will make the necessary changes. Thank for informing!

damithc commented:
Waiting for a clarification.

kennho commented:
Updated with suggested changes @damithc

kennho commented:
Updated the json file to match the latest update @kanghj @damithc
